// css-literal-loader/lib/traverse.js
"use strict";

exports.__esModule = true;
exports.default = traverse;

var _core = require("@babel/core");

var _plugin = _interopRequireDefault(require("./plugin"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function traverse(source, filename, opts) {
  return (0, _core.transformSync)(source, {
    filename,
    babelrc: false,
    code: false,
    ast: false,
    plugins: [[_plugin.default, opts]],
    parserOpts: {
      allowImportExportEverywhere: true,
      allowReturnOutsideFunction: true,
      allowSuperOutsideMethod: true,
      sourceType: 'unambigious',
      sourceFilename: true,
      plugins: ['jsx', 'flow', 'doExpressions', 'objectRestSpread', 'decorators', 'classProperties', 'classPrivateProperties', 'classPrivateMethods', 'exportDefaultFrom', 'exportNamespaceFrom', 'asyncGenerators', 'functionBind', 'functionSent', 'dynamicImport', 'numericSeparator', 'optionalChaining', 'importMeta', 'bigInt', 'optionalCatchBinding', 'throwExpressions', 'pipelineOperator', 'nullishCoalescingOperator']
    }
  });
}
